src/Storage/StreamWrapper.php

```diff
@@ -0,0 +1,278 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Storage;
+
+use Google\Cloud\Exception\GoogleException;
+
+/**
+ * A streamWrapper implementation for handling `gs://bucket/path/to/file.jpg`
+ *
+ * See: http://php.net/manual/en/class.streamwrapper.php
+ */
+class StreamWrapper
+{
+    // Must be public according to the PHP documentation
+    public $context;
+
+    // a GuzzleHttp\Psr7\StreamInterface instance
+    private $stream;
+
+    private $protocol;
+    private $bucket;
+    private $file;
+    private $mode;
+    private $options;
+
+    /**
+     * @var StorageClient $defaultClient The default client to use if using
+     *      global methods such as fopen on a stream wrapper.
+     */
+    private static $client;
+
+    /**
+     * Ensure we close the stream when this StreamWrapper is destroyed.
+     */
+    public function __destruct()
+    {
+        $this->stream_close();
+    }
+
+    /**
+     * Register a StreamWrapper for reading and writing to Google Storage
+     *
+     * @param string $protocol The name of the protocol to use. Defaults to
+     *        'gs'.
+     * @throws \RuntimeException
+     */
+    public static function register(string $protocol = null)
+    {
+        $protocol = $protocol ?: 'gs';
+        if (!in_array($protocol, stream_get_wrappers())) {
+            if (!stream_wrapper_register($protocol, StreamWrapper::class)) {
+                throw new RuntimeException("Failed to register '$protocol://' protocol");
+            }
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Unregisters the SteamWrapper
+     *
+     * @param string $protocol The name of the protocol to unregister. Defaults
+     *        to 'gs'.
+     */
+    public static function unregister(string $protocol = null)
+    {
+        stream_wrapper_unregister($protocol ?: 'gs');
+    }
+
+    /**
+     * Get the default client to use for streams.
+     *
+     * @return StorageClient
+     */
+    public static function getClient()
+    {
+        return self::$client;
+    }
+
+    /**
+     * Set the default client to use for streams.
+     *
+     * @param StorageClient $client
+     */
+    public static function setClient($client)
+    {
+        self::$client = $client;
+    }
+
+    // @codingStandardsIgnoreStart
+    /**
+     * Callback handler for when a stream is opened. For reads, we need to
+     * download the file to see if it can be opened.
+     *
+     * @return bool
+     */
+    public function stream_open($path, $mode, $options, &$opened_path)
+    {
+        // @codingStandardsIgnoreEnd
+        $url = parse_url($path);
+        $this->protocol = $url['scheme'];
+        $this->file = substr($url['path'], 1);
+        $this->mode = $mode;
+
+        $client = $this->getOption('client') ?: self::getClient();
```

Comment:
Yes, that was sort of the point of allowing it. Perhaps it's an edge case that can be added later if it's actually needed. Maybe someone is streaming to/from multiple GCP project configurations?